DOC-4176:  Update IDM configuration documentation (#7)
exoplatform/exo-documentation: docs/LDAP.rst

LDAP Integration
#################

eXo Platform organizational entities (users, groups and memberships),  
can be stored in a database or a directory such as OpenLDAP or Active 
Directory (AD). This chapter documents how to configure eXo Platform to 
plug to a directory.

.. note:: Please notice that this integration is not SSO (Single Sign On).
          If SSO is what you need, read the :ref:`SSO chapter, eXo Add-ons guide <SSO>` that explains how eXo Platform works with a directory through an SSO service like CAS or OpenAM.
    
    
.. warning:: -  eXo Platform supports only the **read-only** mode with a directory (LDAP/AD).
             -  Only one single directory is allowed.
             -  The mapped organizational  entities from directory are imported in one way direction: **from the directory to eXo Platform**.

This chapter covers the following topics:

    -  :ref:`Introduction <LDAP.Introduction>`
       An introduction about directory server integration basics.

    -  :ref:`Quick start <LDAP.QuickStart>`
       A step by step tutorial for eXo Platform configuration with a directory server.

    -  :ref:`Frequently asked questions <LDAP.FAQ>`

>        How to resolve some possible issues of a directory integration.

Reviewer: There is no more "Frequently asked questions" chapter.

.. _LDAP.Introduction:

=============
Introduction
=============

eXo Platform uses `PicketLink IDM framework <http://picketlink.org/>`__ 
that allows a very flexible integration with a directory server:

-  It can be plugged to an already populated directory, in read-only mode. The directory can contain users and groups, or only users.
  
-  Structure of users and groups in the directory can be finely customized.

-  The supported directory implementations are: OpenLDAP and Microsoft Active Directory. You can refer to our official 

>    `supported environments <https://www.exoplatform.com/terms-conditions/supported-environments.pdf>`__ file for more 

Reviewer: ```suggestion
   `supported environments <https://www.exoplatform.com/terms-conditions/supported-environments.pdf>`__ matrix for more 
```

> 

Reviewer: We should remove "The fact that an identity which is created via eXo Platform UI can be saved into the LDAP directory differentiates this integration with SSO use cases." since it is not true anymore.

    If SSO is what you need, read the :ref:`SSO chapter, eXo Add-ons guide <#eXoAddonsGuide.SSO>` that explains how eXo Platform works with
    LDAP through an SSO service like CAS or OpenAM.
    

> .. warning:: -  eXo Platform supports only the **readonly** mode on LDAP/MSAD integration.   

Reviewer: ```suggestion
.. warning:: -  eXo Platform supports only the **read-only** mode on LDAP/MSAD integration.   
```

>              -  Only one single LDAP/MSAD is allowed.

Reviewer: We should unify the term "LDAP/MSAD". Sometimes we use LDAP, sometimes LDAP/AD, sometimes LDAP/MSAD. I would go for LDAP/AD. WDYT ?

Reviewer 2: why not simplifying and saying "directory" instead ?

Reviewer: Sounds good to me.

>              -  The IDM mapped entities from LDAP/MSAD are imported in one way direction: **from LDAP to eXo Platform**. 

Reviewer: ```suggestion
             -  The IDM mapped entities from LDAP/MSAD are imported in one way direction: **from LDAP/MSAD to eXo Platform**. 
```

       A step by step tutorial for eXo Platform configuration with a
       directory server: LDAP/AD.

    -  :ref:`LDAP integration using legacy organization

>        services <LDAP.LegacyOrganizationService>`

Reviewer: These legacy services are not supported anymore, we should remove it from the documentation (I let you choose if you prefer to do it in a dedicated PR or in this one).

Author: in this one

> flexible integration with an LDAP directory:

Reviewer: We should rename this chapter IMO, this can let think that there is another way to integrate with LDAP/AD.
Actually I think that all sub-chapter (Quick start, Configuration review, ...) must be directly put under the main chapter "LDAP Integration" and the content of "LDAP integration using PicketLink" should be moved in a sibling chapter simply called "PicketLink".
WDYT ?

Author: OK

          represents users who are created via eXo Platform UI. The 
          understanding is similar to "*LDAP groups*\ " and "*Platform 

>           groups*\ ".

Reviewer: The following refs (below the note) should be removed since their content is removed:
* Platform users mapped into LDAP
* Platform groups mapped into LDAP

 .. code:: xml

		<attributes>
		...
				   <attribute>
					   <name>user.jobtitle</name>
					   <mapping>title</mapping>
					   <type>text</type>
					   <isRequired>false</isRequired>
					   <isMultivalued>false</isMultivalued>
					   <isUnique>false</isUnique>
				   </attribute>
		...
			   </attributes>
	
			   
.. _LDAP_MultipleDNsGroups:

=====================================
How to map multiple DNs for groups?
=====================================	   
			   
As in previous sections, we assume that you already have a populated directory and some groups that should be mapped into eXo Platform. 

.. tip:: To be clear about the LDAP "group", it should be the "groupOfNames" objectClass in OpenLDAP or "group" objectClass 
         in Active Directory. In OpenLDAP (default core.schema), the groupOfNames must have the member attribute.

Under the context DN (ou=Groups,o=acme,dc=example,dc=com), there are several groups as shown in the diagram below: 

|image2|

In this case, you should, in addition to previous steps described in the :ref:`Quick start section <LDAP.QuickStart>`, 
follow these steps:

1. Open the configuration file ``picketlink-idm-ldap-config.xml``.

> 2. Search for the option ctxDNs to define the different locations of DNs

Reviewer: ```suggestion
2. Search for the option ctxDNs to define the multiple locations of DNs
```

							<value>true</value>
						  </option>
						  <option>
							<name>parentMembershipAttributeName</name>
							<value>member</value>
						  </option>
						  <option>
							<name>isParentMembershipAttributeDN</name>
							<value>true</value>
						  </option>
						  <option>
							<name>allowEmptyMemberships</name>
							<value>true</value>
						  </option>
						  <option>
							<name>createEntryAttributeValues</name>
							<value>objectClass=top</value>
							<value>objectClass=group</value>
							<value>groupType=8</value>
						  </option>
					   </options>
					</identity-object-type>
				</supported-identity-object-types>
			</identity-store>

   Make sure that the attributes and options are correct, especially:
   
   - **idAttributeName**:  attribute name to use as the group id.
   - **ctxDNs**: base DN of the groups in the directory.
   - **entrySearchFilter**: search expression to filter objects to consider as groups.
   - **parentMembershipAttributeName**: attribute which holds the list of group members. In OpenLDAP or MSAD default schemas, 
     the member attribute is used, but your schema may use another attribute.

> Then this new object type must be referenced it in the PortalRepository repository: 

Reviewer: ```suggestion
Then this new object type must be referenced in the PortalRepository repository: 
```

      .. code:: xml
      
			   <repository>
				   <id>PortalRepository</id>
				   ...
				   <identity-store-mapping>
					   <identity-store-id>PortalLDAPStore</identity-store-id>
					   <identity-object-types>
						   ...
						   <identity-object-type>acme_groups_type</identity-object-type>
						   ...
					   </identity-object-types>
				   </identity-store-mapping>...
				</repository>

.. _eXoConfiguration:

2. **eXo configuration**

   Besides the :ref:`PicketLink configuration <PicketlinkConfiguration>`, 

>    the eXo service configuration define in the file ``idm-configuration.xml`` must be updated. 

Reviewer: ```suggestion
   the eXo service configuration defined in the file ``idm-configuration.xml`` must be updated. 
```

   A new entry must be added in the fields ``groupTypeMappings`` and ``ignoreMappedMembershipTypeGroupList`` 

>    to map the group defined in PicketLink configuration with the eXo Platform, as follows: 

Reviewer: ```suggestion
   to map the group defined in PicketLink configuration with the eXo Platform group, as follows: 
```

-  **hibernate.mappings**: list of .xml files that will be added to the hibernate configuration as mapping files.
-  **jndiName** (value-param): if the 'config' parameter is not provided, this parameter will be used to perform the JNDI lookup for IdentitySessionFactory.
-  **portalRealm** (value-param): the realm name that should be used to obtain the proper IdentitySession. The default value is 'PortalRealm'.

   .. code:: xml

			<value-param>
					<name>portalRealm</name>
					<value>idm_realm${container.name.suffix}</value>
			 </value-param>

.. _Ref_eXoConfiguration_PicketLinkIDMOrganizationServiceImpl:

PicketLinkIDMOrganizationServiceImpl service
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This service has the following parameters defined as fields of ``object-param`` 
of type ``org.exoplatform.services.organization.idm.Config``:

-  **rootGroupName** : the name of the PicketLink IDM Group that will be 
   used as a root parent. The default is ``GTN_ROOT_GROUP``.
   
-  **defaultGroupType**: the name of the PicketLink IDM GroupType that 
   will be used to store groups. The default is ``GTN_GROUP_TYPE``.
   

> -  **groupTypeMappings** : this parameter maps groups added with GateIn 

Reviewer: ```suggestion
-  **groupTypeMappings** : this parameter maps groups added with eXo Platform 
```

>    3.5 API as children of a given group ID, and stores them with a given 

Reviewer: ```suggestion
   API as children of a given group ID, and stores them with a given 
```

   group type name in PicketLink IDM.
   If the parent ID ends with "/*", all child groups will have the mapped 
   group type. Otherwise, only direct (first level) children will use this type.
   This can be leveraged by LDAP if the LDAP DN is configured in PicketLink 
   IDM to only store a specific group type. This will then store the given 

>    branch in the GateIn 3.5 group tree, while all other groups will remain in the database.

Reviewer: ```suggestion
   branch in the eXo Platform group tree, while all other groups will remain in the database.
```

   
-  **forceMembershipOfMappedTypes**: groups stored in PicketLink IDM with 
   a type mapped in 'groupTypeMappings' will automatically be members under the mapped parent. 
   The Group relationships linked by the PicketLink IDM group association will not be necessary.

>    This parameter can be set to false if all groups are added via GateIn 3.5 APIs. This may be 

Reviewer: ```suggestion
   This parameter can be set to false if all groups are added via eXo Platform APIs. This may be 
```

   useful with the LDAP configuration when being set to true, it will make every entry added to 

>    LDAP appear in GateIn 3.5. This, however, is not true for entries added via GateIn 3.5 management UI.

Reviewer: ```suggestion
   LDAP appear in eXo Platform. This, however, is not true for entries added via eXo Platform management UI.
```

   
-  **ignoreMappedMembershipType**: if "associationMembershipType" option is used, and this option is set to 
   true, Membership with MembershipType configured to be stored as PicketLink IDM association will not be 
   stored as PicketLink IDM Role.
   
-  **associationMembershipType** : if this option is used, each Membership created with MembrshipType that 
   is equal to the value specified here, will be stored in PicketLink IDM as the simple Group-User association.

-  **passwordAsAttribute**: this parameter specifies if a password should be stored using the PicketLink IDM 
   Credential object or as a plain attribute. The default value is set to false.
   
-  **useParentIdAsGroupType**: this parameter stores the parent ID path as a group type in PicketLink IDM 
   for any IDs not mapped with a specific type in 'groupTypeMappings'. If this option is set to false, 
   and no mappings are provided under 'groupTypeMappings', only one group with the given name can exist 

>    in the GateIn 3.5 group tree.

Reviewer: ```suggestion
   in the eXo Platform group tree.
```

-  OpenLDAP.

-  Microsoft Active Directory.

> -  OpenDS.

Reviewer: Only OpenLDAP and MS AD are supported (at least since PLF 4.4).

    access to * by * read

For more information, refer to `this

> discussion <https://commPlatform users mapped into LDAPunity.exoplatform.com/portal/intranet/forum/topic/topicaf29ef7ca772acc44f16ba9a66b047bf>`__

Reviewer 3: ```suggestion
discussion <https://community.exoplatform.com/portal/intranet/forum/topic/topicaf29ef7ca772acc44f16ba9a66b047bf>`__
```
